src/mempool/block_receiver.rs

```diff
@@ -0,0 +1,190 @@
+            let (block_tx, block_rx) = oneshot::channel::<Option<Block>>();
+            self.system_tx
+                .send(SystemMessage::BlockRequest {
+                    block_event_seqnum: currrent_seqnum,
+                    block_tx,
+                })
+                .await
+                .unwrap();
+            let block = block_rx.await.unwrap().unwrap();
+            self.submit_block(&block).await;
+            if let Some(last_event) = block.events.last() {
+                currrent_seqnum = last_event.seqnum() + 1;
+            }
+        }
+    }
+
+    pub async fn run(&mut self) {
+        info!(shard = self.shard_id.to_string(), "Running block receiver");
+        loop {
+            let block = self.block_rx.recv().await.unwrap();
+            info!(
+                shard = self.shard_id.to_string(),
+                num_events = block.events.len(),
+                height = block.header.as_ref().unwrap().height.unwrap().block_number,
+                "Received block"
+            );
+            if block.events.len() == 0 {
+                continue;
+            }
+            // The db is the source of truth, it's possible to read this out of the events_rx channel but delivery over that channel is not reliable (it's a broadcast channel) we may not have the most up to date state.
+            let last_stored_event_seqnum = self.stores.block_event_store.max_seqnum().unwrap();
+            let last_event_in_block = block.events.last().unwrap();
+            if last_event_in_block.seqnum() < last_stored_event_seqnum {
+                continue;
+            }
+
+            let first_event_in_block = block.events.first().unwrap();
+            if first_event_in_block.seqnum() > last_stored_event_seqnum + 1 {
+                self.sync_missing_block_events(
+                    last_stored_event_seqnum + 1,
+                    first_event_in_block.seqnum() - 1,
+                )
+                .await;
+                if let Err(BlockReceiverError::ConfirmationTimedOut) = self
+                    .wait_for_confirmation(
+                        first_event_in_block.seqnum() - 1,
+                        Duration::from_secs(10),
+                    )
+                    .await
+                {
+                    // TODO(aditi): Right now, we will just wait for the next block with events and try again. In the future we may want better retry logic
+                    warn!(
+                        seqnum = first_event_in_block.seqnum() - 1,
+                        "Timed out waiting for confirmation",
+                    );
+                    continue;
+                }
+            };
+            self.submit_block(&block).await;
+            // If confirmation fails, we'll try move onto the next block and retry this block if needed.
+            if let Err(BlockReceiverError::ConfirmationTimedOut) = self
+                .wait_for_confirmation(last_event_in_block.seqnum(), Duration::from_secs(1))
```

Comment:
These confirmation timeouts should be configurable so they're easy to change. Don't have high conviction in what the numbers should be.